[SPARK-11179] [SQL] Push filters through aggregate

nitin2goyal opened the pull request:
Push conjunctive predicates though Aggregate operators when their references are a subset of the groupingExpressions.

Query plan before optimisation :-
Filter ((c#138L = 2) && (a#0 = 3))
 Aggregate [a#0], [a#0,count(b#1) AS c#138L]
  Project [a#0,b#1]
   LocalRelation [a#0,b#1,c#2]

Query plan after optimisation :-
Filter (c#138L = 2)
 Aggregate [a#0], [a#0,count(b#1) AS c#138L]
  Filter (a#0 = 3)
   Project [a#0,b#1]
    LocalRelation [a#0,b#1,c#2]

nitindexter committed:
SPARK-11179: Push filters through aggregate if filters are subset of 'group by' attribute set

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
Tests please.  Look [here](https://github.com/davies/spark/tree/byte_buffer/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer) for examples.

nitindexter committed:
SPARK-11179: Push filters through aggregate if filters are subset of 'group by' attribute set

nitin2goyal commented:
Added tests

hvanhovell commented:
We could do a similar thing for window functions.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -675,6 +676,20 @@ object PushPredicateThroughGenerate extends Rule[LogicalPlan] with PredicateHelp
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition,
+    aggregate @ Aggregate(groupingExpressions, aggregateExpressions, child))
+        if (filter.condition.references -- AttributeSet(groupingExpressions)).isEmpty =>
```
rather than checking for complete overlap, can we pull out the expressions for group by columns and push those down?

nitin2goyal replied:
Good point. Incorporated.

nitindexter committed:
SPARK-11179: Push filters through aggregate if filters are subset of 'group by' attribute set

marmbrus commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -675,6 +676,29 @@ object PushPredicateThroughGenerate extends Rule[LogicalPlan] with PredicateHelp
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition,
+    aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
```
nit: indentation.  I'm not sure we have a strict rule, but no indent is kinda hard to follow.  I'd probably try to make it a tree if it fits?

```scala
case filter @ Filter(condition,
     aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
```

or just 4 space indent?

marmbrus commented:
This looks great, thanks for doing it!  

Can you cleanup the title `[SPARK-11179] [SQL] Push filters through aggregate`
and the description: `Push conjunctive predicates though Aggregate operators when their references are a subset of the groupingExpressions.`

If you are feeling ambitious I'd include query plans before and after the optimization too.

These become the commit message when we use our merge tool.

SparkQA commented:
**[Test build #43995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43995/consoleFull)** for PR 9167 at commit [`671fbb3`](https://github.com/apache/spark/commit/671fbb31d7c908668526bdc146e0168ffb3014a8).

SparkQA commented:
**[Test build #43995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43995/consoleFull)** for PR 9167 at commit [`671fbb3`](https://github.com/apache/spark/commit/671fbb31d7c908668526bdc146e0168ffb3014a8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43995/
Test FAILed.

nitindexter committed:
[SPARK-11179] [SQL] Push filters through aggregate

Push conjunctive predicates though Aggregate operators when their references are a subset of the groupingExpressions.

Query plan before optimisation :-
Filter ((c#138L = 2) && (a#0 = 3))
 Aggregate [a#0], [a#0,count(b#1) AS c#138L]
  Project [a#0,b#1]
   LocalRelation [a#0,b#1,c#2]

Query plan after optimisation :-
Filter (c#138L = 2)
 Aggregate [a#0], [a#0,count(b#1) AS c#138L]
  Filter (a#0 = 3)
   Project [a#0,b#1]
    LocalRelation [a#0,b#1,c#2]

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

nitin2goyal commented:
Addressed 1 code review comment, fixed couple of scalastyle issues and cleaned up title and description in latest commit and this PR.

SparkQA commented:
**[Test build #44002 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44002/consoleFull)** for PR 9167 at commit [`f422aa8`](https://github.com/apache/spark/commit/f422aa81e10ad01762847c71e678c3b2ef85a926).

SparkQA commented:
**[Test build #44002 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44002/consoleFull)** for PR 9167 at commit [`f422aa8`](https://github.com/apache/spark/commit/f422aa81e10ad01762847c71e678c3b2ef85a926).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44002/
Test FAILed.

nitin2goyal commented:
@marmbrus Test failed again due to whitespace issue. I have already removed whitespace from that line and I am not getting scalastyle issue when compiling locally (don't see any whitespace in code review also). Am I missing something?


[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala:664:0: Whitespace at end of line

marmbrus commented:
try `sbt test:scalastyle`

rxin commented:
You can run local style tests by

dev/lint-scala

nitindexter committed:
[SPARK-11179] [SQL] Push filters through aggregate

Push conjunctive predicates though Aggregate operators when their references are a subset of the groupingExpressions.

Query plan before optimisation :-
Filter ((c#138L = 2) && (a#0 = 3))
Aggregate [a#0], [a#0,count(b#1) AS c#138L]
Project [a#0,b#1]
LocalRelation [a#0,b#1,c#2]

Query plan after optimisation :-
Filter (c#138L = 2)
Aggregate [a#0], [a#0,count(b#1) AS c#138L]
Filter (a#0 = 3)
Project [a#0,b#1]
LocalRelation [a#0,b#1,c#2]

nitin2goyal commented:
Somehow, I wasn't getting scalastyle issue on my branch with above 2 commands. Cloned apache spark master, cherry-picked my changes and then I got the error. Checked-in the whitespace removal. Please test this.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44041/consoleFull)** for PR 9167 at commit [`82fc386`](https://github.com/apache/spark/commit/82fc386675ea2bcd5123d3abd83f6565669fcd69).

SparkQA commented:
**[Test build #44041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44041/consoleFull)** for PR 9167 at commit [`82fc386`](https://github.com/apache/spark/commit/82fc386675ea2bcd5123d3abd83f6565669fcd69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44041/
Test PASSed.

marmbrus commented:
Thanks, merging to master!
